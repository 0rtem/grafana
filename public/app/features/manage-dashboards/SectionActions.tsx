import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import appEvents from '../../core/app_events';
import FormSwitch from 'app/core/components/FormSwitch/FormSwitch';
import { DashboardSectionItem } from '../../types';
import { addTagFilter, setSectionsAndItemsSelected, toggleStarredFilter, loadSections } from './state/actions';
import {
  getAllChecked,
  getCanDelete,
  getCanMove,
  getFilterOnStarred,
  getSelectedDashboards,
  getSelectedTagFilter,
  getTagFilterOptions,
} from './state/selectors';

export interface Props {
  canMove: boolean;
  canDelete: boolean;
  selectedTagFilter: string;
  setSectionsAndItemsSelected: typeof setSectionsAndItemsSelected;
  allChecked: boolean;
  tagFilterOptions: any[];
  addTagFilter: typeof addTagFilter;
  toggleStarredFilter: typeof toggleStarredFilter;
  filterOnStarred: boolean;
  selectedDashboards: DashboardSectionItem[];
  loadSections: typeof loadSections;
}

export class SectionActions extends PureComponent<Props> {
  onSelectAllChanged = () => {
    const { allChecked, setSectionsAndItemsSelected } = this.props;
    setSectionsAndItemsSelected(!allChecked);
  };

  onStarredFilterChange = () => {
    const { filterOnStarred } = this.props;

    this.props.toggleStarredFilter(!filterOnStarred);
  };

  onTagFilterChange = event => {
    this.props.addTagFilter(event.target.value);
  };

  moveTo = () => {
    const { selectedDashboards, loadSections } = this.props;

    const template =
      '<move-to-folder-modal dismiss="dismiss()" ' +
      'dashboards="model.dashboards" after-save="model.afterSave()">' +
      '</move-to-folder-modal>`';

    appEvents.emit('show-modal', {
      templateHtml: template,
      modalClass: 'modal--narrow',
      model: {
        dashboards: selectedDashboards,
        afterSave: loadSections,
      },
    });
  };

  delete = () => {};

  render() {
    const { allChecked, canMove, canDelete, filterOnStarred, selectedTagFilter, tagFilterOptions } = this.props;

    return (
      <div className="search-results">
        <div className="search-results-filter-row">
          <FormSwitch
            label=""
            onChange={this.onSelectAllChanged}
            checked={allChecked}
            switchClass="gf-form-switch--transparent gf-form-switch--search-result-filter-row__checkbox"
          />
          <div>
            {!(canMove || canDelete) && (
              <div className="search-results-filter-row__filters">
                <div style={{ display: 'flex', alignItems: 'center', fontSize: '14px' }}>
                  Filter by starred
                  <FormSwitch
                    label=""
                    checked={filterOnStarred}
                    onChange={this.onStarredFilterChange}
                    switchClass="gf-form-switch--transparent gf-form-switch--search-result-filter-row__checkbox"
                  />
                </div>
                <div className="gf-form-select-wrapper">
                  <select
                    className="search-results-filter-row__filters-item gf-form-input"
                    value={selectedTagFilter}
                    onChange={this.onTagFilterChange}
                  >
                    {tagFilterOptions.map((option, index) => {
                      return <option key={`${option.term}-${index}`}>{option.term}</option>;
                    })}
                  </select>
                </div>
              </div>
            )}
            {(canMove || canDelete) && (
              <div className="gf-form-button-row">
                <button
                  type="button"
                  className="btn gf-form-button btn-inverse"
                  disabled={!canMove}
                  onClick={this.moveTo}
                >
                  <i className="fa fa-exchange" />&nbsp;&nbsp;Move
                </button>
                <button
                  type="button"
                  className="btn gf-form-button btn-danger"
                  disabled={!canDelete}
                  onClick={this.delete}
                >
                  <i className="fa fa-trash" />&nbsp;&nbsp;Delete
                </button>
              </div>
            )}
          </div>
        </div>
      </div>
    );
  }
}

function mapStateToProps(state) {
  return {
    allChecked: getAllChecked(state.sections),
    canDelete: getCanDelete(state),
    canMove: getCanMove(state),
    selectedTagFilter: getSelectedTagFilter(state.manageDashboards),
    tagFilterOptions: getTagFilterOptions(state.sections),
    filterOnStarred: getFilterOnStarred(state.manageDashboards),
    selectedDashboards: getSelectedDashboards(state.sections),
  };
}

const mapDispatchToProps = {
  addTagFilter,
  toggleStarredFilter,
  setSectionsAndItemsSelected,
  loadSections,
};

export default connect(mapStateToProps, mapDispatchToProps)(SectionActions);
