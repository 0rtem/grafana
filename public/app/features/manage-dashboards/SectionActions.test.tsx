import React from 'react';
import { shallow } from 'enzyme';
import { Props, SectionActions } from './SectionActions';

const setup = (propOverrides?: object) => {
  const props: Props = {
    canMove: false,
    canDelete: false,
    selectedTagFilter: '',
    setSectionsAndItemsSelected: jest.fn(),
    allChecked: false,
    tagFilterOptions: [],
    toggleStarredFilter: jest.fn(),
    addTagFilter: jest.fn(),
    filterOnStarred: false,
  };

  Object.assign(props, propOverrides);

  const wrapper = shallow(<SectionActions {...props} />);
  const instance = wrapper.instance() as SectionActions;

  return {
    wrapper,
    instance,
  };
};

describe('Render', () => {
  it('should render component', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should render tag filter options', () => {
    const { wrapper } = setup({
      tagFilterOptions: [
        {
          term: 'mysql',
          count: 1,
        },
        {
          term: 'nosql',
          count: 5,
        },
      ],
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should render move and delete buttons', () => {
    const { wrapper } = setup({
      canMove: true,
      canDelete: true,
    });

    expect(wrapper).toMatchSnapshot();
  });
});

describe('Functions', () => {
  it('should select all', () => {
    const { instance } = setup();

    instance.onSelectAllChanged();

    expect(instance.props.setSectionsAndItemsSelected).toHaveBeenCalledWith(true);
  });

  it('should toggle starred filter', () => {
    const { instance } = setup();

    instance.onStarredFilterChange();

    expect(instance.props.toggleStarredFilter).toHaveBeenCalledWith(true);
  });

  it('should add tag filter', () => {
    const { instance } = setup();
    const mockEvent = { target: { value: 'mysql' } };

    instance.onTagFilterChange(mockEvent);

    expect(instance.props.addTagFilter).toHaveBeenCalledWith('mysql');
  });
});
