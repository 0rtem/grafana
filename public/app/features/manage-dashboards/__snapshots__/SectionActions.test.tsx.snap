// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should render component 1`] = `
<div
  className="search-results"
>
  <div
    className="search-results-filter-row"
  >
    <FormSwitch
      checked={false}
      label=""
      onChange={[Function]}
      switchClass="gf-form-switch--transparent gf-form-switch--search-result-filter-row__checkbox"
    />
    <div>
      <div
        className="search-results-filter-row__filters"
      >
        <div
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "fontSize": "14px",
            }
          }
        >
          Filter by starred
          <FormSwitch
            checked={false}
            label=""
            onChange={[Function]}
            switchClass="gf-form-switch--transparent gf-form-switch--search-result-filter-row__checkbox"
          />
        </div>
        <div
          className="gf-form-select-wrapper"
        >
          <select
            className="search-results-filter-row__filters-item gf-form-input"
            onChange={[Function]}
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Render should render move and delete buttons 1`] = `
<div
  className="search-results"
>
  <div
    className="search-results-filter-row"
  >
    <FormSwitch
      checked={false}
      label=""
      onChange={[Function]}
      switchClass="gf-form-switch--transparent gf-form-switch--search-result-filter-row__checkbox"
    />
    <div>
      <div
        className="gf-form-button-row"
      >
        <button
          className="btn gf-form-button btn-inverse"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <i
            className="fa fa-exchange"
          />
            Move
        </button>
        <button
          className="btn gf-form-button btn-danger"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <i
            className="fa fa-trash"
          />
            Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Render should render tag filter options 1`] = `
<div
  className="search-results"
>
  <div
    className="search-results-filter-row"
  >
    <FormSwitch
      checked={false}
      label=""
      onChange={[Function]}
      switchClass="gf-form-switch--transparent gf-form-switch--search-result-filter-row__checkbox"
    />
    <div>
      <div
        className="search-results-filter-row__filters"
      >
        <div
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "fontSize": "14px",
            }
          }
        >
          Filter by starred
          <FormSwitch
            checked={false}
            label=""
            onChange={[Function]}
            switchClass="gf-form-switch--transparent gf-form-switch--search-result-filter-row__checkbox"
          />
        </div>
        <div
          className="gf-form-select-wrapper"
        >
          <select
            className="search-results-filter-row__filters-item gf-form-input"
            onChange={[Function]}
            value=""
          >
            <option
              key="mysql-0"
            >
              mysql
            </option>
            <option
              key="nosql-1"
            >
              nosql
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
</div>
`;
